core/network-libp2p/src/behaviour.rs

```diff
@@ -293,10 +295,11 @@ impl<TMessage, TSubstream> NetworkBehaviourEventProcess<KademliaOut> for Behavio
 impl<TMessage, TSubstream> NetworkBehaviourEventProcess<PingEvent> for Behaviour<TMessage, TSubstream> {
 	fn inject_event(&mut self, event: PingEvent) {
 		match event {
-			PingEvent::PingSuccess { peer, time } => {
-				trace!(target: "sub-libp2p", "Ping time with {:?}: {:?}", peer, time);
-				self.events.push(BehaviourOut::PingSuccess { peer_id: peer, ping_time: time });
+			PingEvent { peer, result: Ok(PingSuccess::Ping { rtt }) } => {
+				trace!(target: "sub-libp2p", "Ping time with {:?}: {:?}", peer, rtt);
+				self.events.push(BehaviourOut::PingSuccess { peer_id: peer, ping_time: rtt });
 			}
+			_ => ()
```

Comment:
Should we not handle `PingFailure` too?

Reply:
This is just an information for the behaviour. The connection itself closes when a ping fails. This information is redundant with the `Disconnected` event that also contains a message indicating that there was a ping failure.
I wouldn't mind removing `PingFailure` from libp2p entirely.

Reply:
Just to clarify: It depends on the configuration. With the default configuration, the first ping failure causes the connection to be closed. With a higher `num_failures`, all but the last failure are reported as a `PingFailure` events (see [the tests](https://github.com/libp2p/rust-libp2p/blob/master/protocols/ping/src/handler.rs#L308)), which may either be logged and / or incorporated into statistics similarly to `PingSuccess`. Ignoring `PingFailure` just means that, whatever the configuration, all except for the last ping failure is always "silent", which is what happens here and I think is intentional. I still think the event is generally useful in other contexts / configurations but :man_shrugging: